Recipes.Tests/WebApiTelemetryTests.cs

```diff
@@ -116,5 +120,74 @@ public void TearDown()
                            .Should()
                            .Be(new Uri(@"http://contoso.com/"));
         }
+
+        [Test]
+        public async Task Telemetry_events_based_on_HTTP_responses_contain_a_caller_IP_address()
+        {
+            HttpResponseMessage response = null;
+
+            using (Log.With<Telemetry>(t => t.WithPropertiesBasedOn(response)).Enter(() => { }))
+            {
+                var request = new HttpRequestMessage(HttpMethod.Get, @"http://contoso.com/");
+                request.Properties[RemoteEndpointMessageProperty.Name] = new RemoteEndpointMessageProperty("123.123.123.123", 80);
+
+                response = new HttpResponseMessage(HttpStatusCode.OK)
+                {
+                    RequestMessage = request
+                };
+            }
+
+            telemetryEvents.Single()
+                           .IsIncomingRequest()
+                           .Should()
+                           .BeTrue();
+            telemetryEvents.Single()
+                           .CallerIpAddress()
+                           .Should()
+                           .Be("123.123.123.123");
+        }
+
+        [Test]
+        public async Task Telemetry_events_based_on_HTTP_responses_do_not_contain_an_action_descriptor()
+        {
+            HttpResponseMessage response = null;
+
+            using (Log.With<Telemetry>(t => t.WithPropertiesBasedOn(response)).Enter(() => { }))
+            {
+                response = new HttpResponseMessage(HttpStatusCode.OK)
+                {
+                    RequestMessage = new HttpRequestMessage(HttpMethod.Get, @"http://contoso.com/operationName5")
+                };
+            }
+
+            telemetryEvents.Single()
+                           .OperationName
+                           .Should()
+                           .Be("operationName5");
+        }
+
+        [Test]
+        public async Task Telemetry_events_based_on_HTTP_responses_contain_an_action_descriptor()
+        {

```

Comment:
This name doesn't quite get to the heart of the issue. It's the request, not the response, that has an ActionDescriptor. Also, this will only apply when hosted under an actual web server, if I recall correctly. In in-memory mode, URL parsing is used instead. The test name should maybe clarify that.